NMS-9146: Added fix for changing location to blank value

christianpape opened the pull request:
* JIRA: https://issues.opennms.org/browse/NMS-9146

mvrueden commented on opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/DefaultProvisionService.java:
```diff
@@ -789,6 +791,10 @@ public OnmsNode getRequisitionedNode(final String foreignSource, final String fo
         node.setCategories(dbCategories);
 
+        if (node.getLocation() == null || Strings.isNullOrEmpty(node.getLocation().getLocationName())) {
+            node.setLocation(m_monitoringLocationDao.getDefaultLocation());
```
Why is this still needed? Shouldn't the `SaveOrUpdateOperation` already cover this?

christianpape replied:
No, this is required by the NodeScan stuff. It uses this method to get the node from the requisition.

mvrueden commented on opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/operations/SaveOrUpdateOperation.java:
```diff
@@ -59,7 +62,13 @@
+        m_node = new OnmsNode();
+
+        m_node.setLocation(Strings.isNullOrEmpty(location) ?
+                new OnmsMonitoringLocation(MonitoringLocationDao.DEFAULT_MONITORING_LOCATION_ID, MonitoringLocationDao.DEFAULT_MONITORING_LOCATION_ID) :
```
Can we not set it to `null` here instead? The `updateLocation` from `ProvisionService.insertNode()` or `ProvisionerService.updateNode()` should handle it accordingly then, shouldn't it?

christianpape replied:
The HostnameResolver stuff in the same class throws a NPE when setting to the location to null.

pioto commented:
Could a migration be added to correct any existing bad data from this bug?

mvrueden commented:
Adding a Upgradetask definitely makes sense. 👍

j-white commented:
This should also affect a small number of users, so documentation on how to remedy the problem in the release notes should be sufficient.

Writing an upgrade task will be tricky, since we would need to alter both the database and the requisitions.

christianpape commented:
Yes, I agree. A update-task makes no sense. I think re-synchronizing the affected requisitions with blank location entries is sufficient. The only thing left to do is to delete the monitoring location with the blank name. This should be possible in the admin's UI, right?

pioto commented:
Doing that seems to delete ALL monitoring locations, and all nodes with
them.

On Fri, Mar 3, 2017, 1:57 AM Christian Pape <notifications@github.com>
wrote:

> Yes, I agree. A update-task makes no sense. I think re-synchronizing the
> affected requisitions with blank location entries is sufficient. The only
> thing left to do is to delete the monitoring location with the blank name.
> This should be possible in the admin's UI, right?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenNMS/opennms/pull/1355#issuecomment-283880967>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AABA4sOrOKz1O-ZQ0f6QEjxyLYI528Ewks5rh7nMgaJpZM4MQ2Cv>
> .
>
-- 

Mike Kelly

mvrueden commented:
@pioto you have to first resync the requisition and delete it afterwards I would imagine.

christianpape commented:
Yes, exactly.

pioto commented:
I think you misunderstand the issue -- if you delete the `""` location, the code seems to behave as if you ran:

```sql
DELETE FROM monitoringlocations
```

It deletes all monitoring locations (including the `"Default"` one), and with them, all nodes...

christianpape commented:
Ok. Sorry, now I got it. I fixed the requisition/provisioning stuff to correctly handle blank locations. The deletion of all nodes is due to the fact, that with an blank location passed to the ReST service the deleteMany method is called. This is totally unrelated to the provisioning stuff. With my fixes you cannot end up with an blank location but of course this will not help if you already have one... I'll update this PR.

christianpape commented:
I updated this PR:
* added liquibase change set to assure the Default location exists
* added liquibase change set to change references to the blank location to the Default location
* added liquibase change set to remove the ON DELETE CASCADE constraint for the node table
* added liquibase change set to remove the blank location
* added error message to "Manage Monitoring Locations" UI

@mvrueden Can you please review and test again?

mvrueden commented on core/schema/src/main/liquibase/19.0.2/changelog.xml:
```diff
@@ -0,0 +1,145 @@
+          )
+          then 1
+          else 0
+        end;
```
Can`t we just do something simpler like `select count(location) from node where location=''`.

christianpape replied:
Sorry, you can only compare expectedResult to a single value and cannot create things like 'greater than' or 'lower than'. So, the recommended workaround is to use these 'case' statements.

mvrueden commented on core/schema/src/main/liquibase/19.0.2/changelog.xml:
```diff
@@ -0,0 +1,145 @@
+          )
+          then 1
+          else 0
+        end;
```
See above.
